pydicer/input/orthanc.py

```diff
@@ -0,0 +1,135 @@
+
+from pydicer.input.base import InputBase
+
+logger = logging.getLogger(__name__)
+
+
+def adapt_dataset_from_bytes(blob):
+    """Convert bytes coming from Orthanc to DICOM dataset
+
+    Args:
+        blob (bytes): The bytes to convert
+
+    Returns:
+        pydicom.Dataset: The DICOM dataset
+    """
+    dataset = pydicom.dcmread(BytesIO(blob))
+    return dataset
+
+
+class OrthancInput(InputBase):
+    def __init__(self, host, port, username=None, password=None, working_directory=None):
+        """Class for fetching files from Orthanc.
+
+        Args:
+            host (str): The IP address or host name of the Orthanc.
+            port (int): The port to use to communicate on.
+            username (str, optional): Orthanc username.
+            password (str, optional): Orthanc password.
+            working_directory (str|pathlib.Path, optional): The working directory in which to
+            store the data fetched. Defaults to a temp directory.
+
+        Raises:
+            ConnectionError: Raises a connection error if unable to verify the connection to
+                Orthanc.
+        """
+
+        super().__init__(working_directory)
+
+        if not host.startswith("http"):
+            host = f"http://{host}"
+
+        self.orthanc = Orthanc(f"{host}:{port}")
+
+        if username is not None and password is not None:
+            self.orthanc.setup_credentials(username, password)
+
+        # Do a dummy lookup to check that we can reach the Orthanc host, this will throw a
+        # connection error if we can't connect to the Orthanc
+        self.orthanc.c_find({"Level": "Patient", "Query": {"PatientID": "XXX"}})
+
+    def fetch_data(self, patients, modalities=None):
+        """Download the DICOM data from Orthanc
+
+        Args:
+            patients (list|str): A list of patient IDs, or a single patient ID.
+            modalities (list|str, optional): List of modalities or a single modality to fetch.
+                Defaults to None where all modalities would be fetched.
+        """
+
+        if not isinstance(patients, list) and not isinstance(patients, tuple):
+            patients = [patients]
+
+        if (
+            modalities is not None
+            and not isinstance(modalities, list)
+            and not isinstance(modalities, tuple)
+        ):
+            modalities = [modalities]
+
+        for patient in patients:
+
+            # Find the Orthanc ID for this patient
+            orthanc_patient_ids = self.orthanc.c_find(
+                {"Level": "Patient", "Query": {"PatientID": patient}}
+            )
+
+            if len(orthanc_patient_ids) == 0:
+                logger.warning("Patient not found in Orthanc: %s", patient)
+
+            if len(orthanc_patient_ids) > 1:
+                logger.warning(
+                    "Patient returned multple Orthanc IDs: %s. Selecting first only", patient
+                )
+
+            orthanc_patient_id = orthanc_patient_ids[0]
```

Comment:
It's oblivious to me at the moment, but there is any reason why we can't do it for all patients returned by orthanc?

Reply:
Oh I see know, we iterate over the list of patients, then do it singularly

Reply:
Yep it loops over each patient passed into this function. But orthanc has a different patient_id scheme so we need to look that up first... That should always return 1 id from Orthanc (or zero if they aren't in there). But I log a warning in case more than one comes back for some reason but really should never happen...

There might be a way to query Orthanc with all the patient ids which we want, but Im not sure how to do that or if its possible...